src/java.base/share/classes/sun/util/locale/provider/BreakIteratorProviderImpl.java

```diff
@@ -193,4 +195,121 @@ public Set<String> getAvailableLanguageTags() {
     public boolean isSupportedLocale(Locale locale) {
         return LocaleProviderAdapter.forType(type).isSupportedProviderLocale(locale, langtags);
     }
+
+    static final class GraphemeBreakIterator extends BreakIterator {
+        private CharacterIterator ci;
+        private int offset;
+        private List<Integer> boundaries;
+        private int boundaryIndex;
+
+        GraphemeBreakIterator() {
+            boundaries = Collections.emptyList();
+            boundaryIndex = 0;
+            offset = 0;
+        }
+
+        @Override
+        public int first() {
+            boundaryIndex = 0;
+            return current();
+        }
+
+        @Override
+        public int last() {
+            boundaryIndex = boundaries.size() - 1;
+            return current();
+        }
+
+        @Override
+        public int next(int n) {
+            if (n == 0) {
+                return offset;
+            }
+
+            var oldIndex = boundaryIndex;
+            boundaryIndex = boundaryIndex + n <= 0 ? 0 : Math.min(boundaryIndex + n, boundaries.size() - 1);
+
+            if (oldIndex == boundaryIndex && offset == boundaries.get(boundaryIndex)) {
+                return DONE;
+            } else {
+                return current();
+            }
+        }
+
+        @Override
+        public int next() {
+            return next(1);
+        }
+
+        @Override
+        public int previous() {
+            return next(-1);
+        }
+
+        @Override
+        public int following(int offset) {
+            if (offset >= this.offset && this.offset == boundaries.get(boundaries.size() - 1)) {
+                return DONE;
+            }
+
+            boundaryIndex = IntStream.range(0, boundaries.size())
+                    .filter(i -> boundaries.get(i) > offset)
+                    .findFirst()
+                    .orElse(boundaries.size() - 1);
```

Comment:
Is it worth trying to use Collections.binarySearch() here? I think the boundaries list is in ascending sorted order, so you might be able to drop in a binarySearch() call directly. (Need to be a bit careful with negative return values.)

Reply:
Will replace this with `Collections.binarySearch()`.